### PR TITLE
Add Windows split tunneling support to the daemon

### PR DESCRIPTION
**Pull request by dlon**
Using the [split tunnel driver](https://github.com/mullvad/win-split-tunnel). GUI frontend changes are separate. This also does not include any changes to the installer (#2666).

---
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)

### CODE REVIEW

**Review by mvd-ows — COMMENTED** [review]
> Reviewed 14 of 41 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 14 of 41 files reviewed, 1 unresolved discussion (waiting on @dlon)

___
*[windows/winnet/src/winnet/winnet.cpp, line 527 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_2bhHeAp5ynospqcRc:-M_2bhHf4it4Pwhm7wtr:b32ffgg) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/windows/winnet/src/winnet/winnet.cpp#L527)):*
> ```cpp
> 	if (nullptr == g_RouteManager)
> 	{
> 		if (nullptr != g_RouteManagerLogSink)
> ```

I'm surprised that `WinNet_DeactivateRouteManager()` doesn't reset `g_RouteManagerLogSink`. It probably should.

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> Reviewed 4 of 41 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 25 of 41 files reviewed, 3 unresolved discussions (waiting on @dlon)

___
*[mullvad-management-interface/src/types.rs, line 368 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_5NANo4W7ltxtoXs5P:-M_5NANp1syf0aGt3C5O:bz84kug) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/mullvad-management-interface/src/types.rs#L368)):*
> ```XML
>             show_beta_releases: settings.show_beta_releases,
>             #[cfg(windows)]
>             split_tunnel: settings.split_tunnel,
> ```

Maybe the windows specific split tunneling fields belong in a windows specific struct?
___
*[mullvad-types/src/settings/mod.rs, line 90 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_5NSUJ4KPlhn37d0mR:-M_5NSUJ4KPlhn37d0mS:b7p00q) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/mullvad-types/src/settings/mod.rs#L90)):*
> ```XML
>             split_tunnel: false,
>             #[cfg(windows)]
>             split_tunnel_apps: HashSet::new(),
> ```

Maybe keep these fields in a separate Windows specific split tunnel struct?
___
*[talpid-core/src/winnet.rs, line 141 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_5O9uB4FKRDE5uZU2u:-M_5O9uB4FKRDE5uZU2v:bh2bzex) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/talpid-core/src/winnet.rs#L141)):*
> ```XML
>     fn from(addr: WinNetIp) -> Ipv6Addr {
>         match addr.addr_family {
>             WinNetAddrFamily::IPV4 => panic!("address family mismatch"),
> ```

I don't think `From` traits should ever panic. Could this not be a `TryFrom` trait instead? Same goes for the `Ipv4Addr` trait.

**Review by dlon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 20 of 41 files reviewed, 2 unresolved discussions (waiting on @dlon and @pinkisemils)

___
*[mullvad-daemon/src/management_interface.rs, line 686 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_0yfMr0B38HCCawzVl:-M_69Fd76QWU88Co6wNU:b-64zfqb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/mullvad-daemon/src/management_interface.rs#L686)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Not sure about this.
</blockquote></details>

I've removed this RPC.
___
*[mullvad-management-interface/src/types.rs, line 368 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_5NANo4W7ltxtoXs5P:-M_69O4N8XtyZzp8j8As:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/mullvad-management-interface/src/types.rs#L368)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Maybe the windows specific split tunneling fields belong in a windows specific struct?
</blockquote></details>

Done.
___
*[talpid-core/src/winnet.rs, line 141 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_5O9uB4FKRDE5uZU2u:-M_69Vap6hnDv3U8RpNH:bjeg83v) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/talpid-core/src/winnet.rs#L141)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

I don't think `From` traits should ever panic. Could this not be a `TryFrom` trait instead? Same goes for the `Ipv4Addr` trait.
</blockquote></details>

Done. `From<WinNetIp> for IpAddr` still unwraps now, but it is impossible for it to panic.

**Review by dlon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 19 of 41 files reviewed, 1 unresolved discussion (waiting on @mvd-ows and @pinkisemils)

___
*[windows/winnet/src/winnet/winnet.cpp, line 527 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_2bhHeAp5ynospqcRc:-M_6Cu9r7h7C-KIh1eLF:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/windows/winnet/src/winnet/winnet.cpp#L527)):*
<details><summary><i>Previously, mvd-ows wrote…</i></summary><blockquote>

I'm surprised that `WinNet_DeactivateRouteManager()` doesn't reset `g_RouteManagerLogSink`. It probably should.
</blockquote></details>

Done.

**Review by mvd-ows — COMMENTED** [review]
> Reviewed 1 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 20 of 41 files reviewed, 1 unresolved discussion (waiting on @pinkisemils)

**Review by dlon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 37 of 41 files reviewed, 3 unresolved discussions (waiting on @pinkisemils)

___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 165 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_5PRR80QPkgPQLM0-m:-M_KkOWJ4Ldnkno9xXUX:b2y0dfr) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/10187a700ab902b8b16398fe339fe254c83318ef/talpid-core/src/split_tunnel/windows/driver.rs#L165)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

This might be helpful during development, but do we need to get the state again here just for logging?
</blockquote></details>

I've removed all of these.
___
*[talpid-core/src/tunnel_state_machine/mod.rs, line 52 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KbXkc-x4ZDgDqAoST:-M_KkcDK1GWIgu_ojJl9:b-hsk1cw) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9068494b2daf4faca5977fc3f0f531a7814c9d27/talpid-core/src/tunnel_state_machine/mod.rs#L52)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Is this error no longer relevant on Linux?
</blockquote></details>

Nope. `PidManager` is not at all dependent on the tunnel state machine, so it is set up in `mullvad-daemon` instead. This must have been forgotten about.
___
*[talpid-core/src/tunnel_state_machine/mod.rs, line 137 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_6YloDBkB1h9g6YtIe:-M_Klw3E34bP5eAqg3-p:bq2kn4w) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/tunnel_state_machine/mod.rs#L137)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Could this not be `command_tx.clone()`? 
</blockquote></details>

No, because that will move `command_tx`.
___
*[talpid-core/src/tunnel_state_machine/mod.rs, line 234 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_6YsfbCDoBPdJLyz4Y:-M_Kmzk43gtolhPM49yq:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/tunnel_state_machine/mod.rs#L234)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

I think the 2 new arguments should be appended to the end of the arg list, without `reset_firewall` in between.
</blockquote></details>

Done.

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 37 of 41 files reviewed, 6 unresolved discussions (waiting on @dlon and @pinkisemils)

___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 351 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KgbfR1mBj3lYUWwEg:-M_KgbfR1mBj3lYUWwEh:b-5eb2bb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L351)):*
> ```XML
> }
> 
> /// Create a buffer containing information.
> ```

This is a very general statement.
___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 357 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KisXRAjAFV4P2Sqbj:-M_KisXRAjAFV4P2Sqbk:b-4oqs34) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L357)):*
> ```XML
>     let apps: Vec<Vec<u16>> = apps
>         .iter()
>         .map(|app| app.as_ref().encode_wide().collect())
> ```

`OsStr` implements [`encode_wide`](https://doc.rust-lang.org/std/os/windows/ffi/trait.OsStrExt.html#tymethod.encode_wide) which is an iterator for `u16`. 
___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 391 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KhK9WAR6r6C0zY-Da:-M_KhK9WAR6r6C0zY-Db:bjcy3mc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L391)):*
> ```XML
>     let string_data = unsafe {
>         std::slice::from_raw_parts_mut(
>             &mut buffer[(total_buffer_size - total_string_size)..] as *mut _ as *mut u16,
> ```

Shouldn't this subslice take the `size_of::<ConfigurationHeader>()` offset too?
___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 398 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_Kj4za0p4_m3AH4-hS:-M_Kj4za0p4_m3AH4-hT:bqwsxhm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L398)):*
> ```XML
> 
>     for (i, app) in apps.iter().enumerate() {
>         string_data[string_offset..string_offset + app.len()].copy_from_slice(app);
> ```

I'd personally just have a single prefer to have a single slice to write to rather than have two slices that point to the same vector. But that might be more complex. Feel free to leave this as is, as this kind of code is rewritten a lot below.

**Review by pinkisemils — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 37 of 41 files reviewed, 3 unresolved discussions (waiting on @dlon)

___
*[talpid-core/src/tunnel_state_machine/mod.rs, line 137 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_6YloDBkB1h9g6YtIe:-M_KveRe2jGvId8Lwc6j:bmsk75v) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/tunnel_state_machine/mod.rs#L137)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

No, because that will move `command_tx`.
</blockquote></details>

Ah yes, the venerable `move || {}` closure :/

**Review by dlon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 37 of 41 files reviewed, 3 unresolved discussions (waiting on @dlon and @pinkisemils)

___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 357 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KisXRAjAFV4P2Sqbj:-M_L-zxv3Ipl4Czdkb7F:b4nzvic) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L357)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

`OsStr` implements [`encode_wide`](https://doc.rust-lang.org/std/os/windows/ffi/trait.OsStrExt.html#tymethod.encode_wide) which is an iterator for `u16`. 
</blockquote></details>

I think I was concerned about `OsStr::len` not matching the length of the `u16` string, since one character can be encoded as multiple `u16` code units. Meaning the lengths would have to be counted twice.

Maybe rather than determining the buffer size in advance, strings could be stored in a temporary `Vec<u16>`, or appended directly to `buffer`? Although the latter feels like unnecessary optimization.

**Review by dlon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 37 of 41 files reviewed, 3 unresolved discussions (waiting on @pinkisemils)

___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 351 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KgbfR1mBj3lYUWwEg:-M_LKC7t4SHN5PLOQaWC:bt2vr1b) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L351)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

This is a very general statement.
</blockquote></details>

Fixed.
___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 357 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KisXRAjAFV4P2Sqbj:-M_LKeCUA4BHP_P0RdIs:b-ybxfw9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L357)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

I think I was concerned about `OsStr::len` not matching the length of the `u16` string, since one character can be encoded as multiple `u16` code units. Meaning the lengths would have to be counted twice.

Maybe rather than determining the buffer size in advance, strings could be stored in a temporary `Vec<u16>`, or appended directly to `buffer`? Although the latter feels like unnecessary optimization.
</blockquote></details>

Actually, I also did it this way to avoid resizing the buffer and invalidate the slices/pointers. So I think this is fine.
___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 391 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_KhK9WAR6r6C0zY-Da:-M_LKHDL22RwIUfiNfuC:brnw2cb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/24b35cd1ebeb30ad7f4c63eabbf829adde3b281a/talpid-core/src/split_tunnel/windows/driver.rs#L391)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Shouldn't this subslice take the `size_of::<ConfigurationHeader>()` offset too?
</blockquote></details>

The offset equals `sizeof(header) + numEntries * sizeof(entries)`, which is correct. The strings are preceded by all the entries.

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> Reviewed 1 of 10 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: 39 of 41 files reviewed, 5 unresolved discussions (waiting on @dlon and @pinkisemils)

___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 788 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_L2Wci0IRYHA8sNv_v:-M_L2Wci0IRYHA8sNv_w:b-fj4rox) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7787bbaec11b415e72577cdf76f2d51c9fa9961a/talpid-core/src/split_tunnel/windows/driver.rs#L788)):*
> ```XML
> 
>     if result != 0 {
>         // This should not occur
> ```

If this shouldn't occur, this should definitely be logged. Or would the error be unique enough to signal to the caller that this was something that should never happen?
___
*[talpid-core/src/split_tunnel/windows/driver.rs, line 863 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726#-M_LJOYr0CwvGZ1XumHY:-M_LJOYr0CwvGZ1XumHZ:b2nreh6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/7787bbaec11b415e72577cdf76f2d51c9fa9961a/talpid-core/src/split_tunnel/windows/driver.rs#L863)):*
> ```XML
>     if result != 0 {
>         // This should not occur
>         return Err(io::Error::last_os_error());
> ```

Same as above.

**Review by pinkisemils — APPROVED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2726)* status: :shipit: complete! all files reviewed, all discussions resolved

File: talpid-core/src/split_tunnel/windows/mod.rs [hunk 1]
```diff
@@ -467,30 +467,34 @@ unsafe extern "system" fn split_tunnel_default_route_change_handler(
...
+                    IpAddr::V6(addr) => ctx.internet_ipv6 = Some(addr),
+                },
                 Ok(None) => {
-                    log::error!("Failed to obtain new default route address: none found",);
-                    maybe_send(TunnelCommand::Block(ErrorStateCause::SplitTunnelError));
-                    return;
+                    log::warn!("Failed to obtain default route interface address");
+                    match address_family {
+                        WinNetAddrFamily::IPV4 => {
+                            ctx.internet_ipv4 = Ipv4Addr::new(0, 0, 0, 0);
```
Comment (mvd-ows) [inline talpid-core/src/split_tunnel/windows/mod.rs:479]: Why isn't this using `Option<Ipv4Addr>`/`None`?

File: talpid-core/src/split_tunnel/windows/mod.rs [hunk 2]
```diff
@@ -346,34 +347,28 @@ impl SplitTunnel {
...
         self._route_change_callback = None;
 
-        Self::register_ips(
-            &*self.handle,
-            Ipv4Addr::new(0, 0, 0, 0),
-            None,
-            Ipv4Addr::new(0, 0, 0, 0),
-            None,
-        )
+        Self::register_ips(&*self.handle, Ipv4Addr::new(0, 0, 0, 0), None, None, None)
```
Comment (mvd-ows) [inline talpid-core/src/split_tunnel/windows/mod.rs:350]: Second argument can also be `None`?

**Review by mvd-ows — APPROVED** [review]

File: talpid-core/src/split_tunnel/windows/mod.rs (hunk 2 above)
Comment (dlon) [reply talpid-core/src/split_tunnel/windows/mod.rs:350]: It probably should be. Anyway, this can be improved later. Leaving it like this for now.

**Pull request merged**